pkg/nodeupgrader/upgrader.go

```diff
@@ -0,0 +1,144 @@
+package nodeupgrader
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/utils/ptr"
+)
+
+const (
+	upgradeScript = "/foo/eksa-upgrades/scripts/upgrade.sh"
+
+	// CopierContainerName holds the name of the components copier container.
+	CopierContainerName = "components-copier"
+
+	// ContainerdUpgraderContainerName holds the name of the containerd upgrader container.
+	ContainerdUpgraderContainerName = "containerd-upgrader"
+
+	// CNIPluginsUpgraderContainerName holds the name of the CNI plugins upgrader container.
+	CNIPluginsUpgraderContainerName = "cni-plugins-upgrader"
+
+	// KubeadmUpgraderContainerName holds the name of the kubeadm upgrader container.
+	KubeadmUpgraderContainerName = "kubeadm-upgrader"
+
+	// KubeletUpgradeContainerName holds the name of the kubelet/kubectl upgrader container.
+	KubeletUpgradeContainerName = "kubelet-kubectl-upgrader"
+
+	// PostUpgradeContainerName holds the name of the post upgrade cleanup/status report container.
+	PostUpgradeContainerName = "post-upgrade-status"
+)
+
+// PodName returns the name of the upgrader pod based on the nodeName.
+func PodName(nodeName string) string {
+	return fmt.Sprintf("%s-node-upgrader", nodeName)
+}
+
+// UpgradeFirstControlPlanePod returns an upgrader pod that should be deployed on the first control plane node.
+func UpgradeFirstControlPlanePod(nodeName, image, kubernetesVersion, etcdVersion string) *corev1.Pod {
+	p := upgraderPod(nodeName, image)
+	p.Spec.InitContainers = containersForUpgrade(image, nodeName, "kubeadm_in_first_cp", kubernetesVersion, etcdVersion)
+	return p
+}
+
+// UpgradeRestControlPlanePod returns an upgrader pod that can be deployed on the remaining control plane nodes.
+func UpgradeRestControlPlanePod(nodeName, image string) *corev1.Pod {
+	p := upgraderPod(nodeName, image)
+	p.Spec.InitContainers = containersForUpgrade(image, nodeName, "kubeadm_in_rest_cp")
+	return p
+}
+
+// UpgradeWorkerPod returns an upgrader pod that can be deployed on worker nodes.
+func UpgradeWorkerPod(nodeName, image string) *corev1.Pod {
```

Comment:
These feel like they should be member methods

Reply:
hmm, let's discuss this offline and see if this is something we can acknowledge in a follow up